python/ql/test/experimental/dataflow/match/test.py

```diff
@@ -0,0 +1,152 @@
+
+import sys
+import os
+
+sys.path.append(os.path.dirname(os.path.dirname((__file__))))
+from testlib import *
+
+# These are defined so that we can evaluate the test code.
+NONSOURCE = "not a source"
+SOURCE = "source"
+
+
+def is_source(x):
+    return x == "source" or x == b"source" or x == 42 or x == 42.0 or x == 42j
+
+
+def SINK(x):
+    if is_source(x):
+        print("OK")
+    else:
+        print("Unexpected flow", x)
+
+
+def SINK_F(x):
+    if is_source(x):
+        print("Unexpected flow", x)
+    else:
+        print("OK")
+
+def test_guard():
+    match SOURCE:
+        case x if SINK(x): #$ flow="SOURCE, l:-1 -> x"
+            pass
+
+@expects(2)
+def test_as_pattern():
+    match SOURCE:
+        case x as y:
+            SINK(x) #$ flow="SOURCE, l:-2 -> x"
+            SINK(y) #$ flow="SOURCE, l:-3 -> y"
+
+def test_or_pattern():
+    match SOURCE:
+        # We cannot use NONSOURCE in place of "" below, since it would be seen as a variable.
+        case ("" as x) | x:
+            SINK(x) #$ flow="SOURCE, l:-3 -> x"
+
+# No flow for literal pattern
+
+def test_capture_pattern():
+    match SOURCE:
+        case x:
+            SINK(x) #$ flow="SOURCE, l:-2 -> x"
+
+# No flow for wildcard pattern
+
+class Unsafe:
+    VALUE = SOURCE
+
+def test_value_pattern():
+    match SOURCE:
+        case Unsafe.VALUE as x:
+            SINK(x) #$ flow="SOURCE, l:-2 -> x" MISSING: flow="SOURCE, l:-5 -> x"
+
+@expects(2)
+def test_sequence_pattern_tuple():
+    match (NONSOURCE, SOURCE):
+        case (x, y):
+            SINK_F(x)
+            SINK(y) #$ flow="SOURCE, l:-3 -> y"
+
+@expects(2)
+def test_sequence_pattern_list():
+    match [NONSOURCE, SOURCE]:
+        case [x, y]:
+            SINK_F(x) #$ SPURIOUS: flow="SOURCE, l:-2 -> x"
```

Comment:
I note the spurious flow here, which I assume is caused by the list content mentioned elsewhere. I feel like we should be able to tighten this up a bit.

Reply:
In this case, the incoming subject `[NONSOURCE, SOURCE]` has list content, so we really do not know where in the list the taint resides. In more realistic code, it would be written as
```
match l:
  case [x, y]:
    ... use x ...
```
If `l` has list content, we have to read it into `x`.